File: db-4.8.30/java/src/com/sleepycat/db/CompactConfig.java
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2002-2009 Oracle.  All rights reserved.
 *
 * $Id$
 */

package com.sleepycat.db;

import com.sleepycat.db.internal.DbConstants;

/**
Configuration for {@link Database#compact} operations.
*/
public class CompactConfig implements Cloneable {
    /**
    Default configuration used if null is passed to methods that create a
    cursor.
    */
    public static final CompactConfig DEFAULT = new CompactConfig();

    /* package */
    static CompactConfig checkNull(CompactConfig config) {
        return (config == null) ? DEFAULT : config;
    }

    private int fillpercent = 0;
    private boolean freeListOnly = false;
    private boolean freeSpace = false;
    private int maxPages = 0;
    private int timeout = 0;

    /**
    Construct a default configuration object for compact operations.
    */
    public CompactConfig() {
    }

    /**
    Set the desired fill percentage.
    If non-zero, the goal for filling pages, specified as a percentage
    between 1 and 100.  Any page in a Btree or Recno databases not at or
    above this percentage full will be considered for compaction.  The
    default behavior is to consider every page for compaction, regardless
    of its page fill percentage.
    @param fillpercent
    The desired fill percentage.
    **/
    public void setFillPercent(final int fillpercent) {
        this.fillpercent = fillpercent;
    }

    /**
Return the the desired fill percentage.
<p>
This method may be called at any time during the life of the application.
<p>
@return
The the desired fill percentage.
 */
    public int getFillPercent() {
        return fillpercent;
    }

    /**
    Configure whether to skip page compaction, only returning pages
    to the filesystem that are already free and at the end of the file.
    This flag must be set if the database is a Hash access method database.
    @param freeListOnly
    Whether to skip page compaction
   **/
    public void setFreeListOnly(boolean freeListOnly) {
        this.freeListOnly = freeListOnly;
    }

    /**
Return true if the whether to skip page compaction.
<p>
This method may be called at any time during the life of the application.
<p>
@return
True if the whether to skip page compaction.
 */
    public boolean getFreeListOnly() {
        return freeListOnly;
    }

    /**
    Return pages to the filesystem if possible.  If this flag is not
    specified, pages emptied as a result of compaction will be placed on the
    free list for re-use, but not returned to the filesystem.
    Note that only pages at the end of the file may be returned.  Given the one
    pass nature of the algorithm if a page near the end of the file is
    logically near the begining of the btree it will inhibit returning pages to
    the file system.
    A second call to the method with a low fillfactor can be used to return
    pages in such a situation.
    @param freeSpace
    Whether to return pages to the filesystem
    **/
    public void setFreeSpace(boolean freeSpace) {
        this.freeSpace = freeSpace;
    }

    /**
Return true if the whether to return pages to the filesystem.
<p>
This method may be called at any time during the life of the application.
<p>
@return
True if the whether to return pages to the filesystem.
 */
    public boolean getFreeSpace() {
        return freeSpace;
    }

    /**
    Set the maximum number of pages to free.
    @param maxPages
    If non-zero, the call will return after that number of pages have been
    freed.
    **/
    public void setMaxPages(final int maxPages) {
        this.maxPages = maxPages;
    }

    /**
Return the the maximum number of pages to free.
<p>
This method may be called at any time during the life of the application.
<p>
@return
The the maximum number of pages to free.
 */
    public int getMaxPages() {
        return maxPages;
    }

    /**
    Set the lock timeout for implicit transactions.
    If non-zero, and no transaction parameter was specified to
    {@link Database#compact}, the lock timeout set for implicit
    transactions, in microseconds.
    @param timeout
    the lock timeout set for implicit transactions, in microseconds.
    **/
    public void setTimeout(final int timeout) {
        this.timeout = timeout;
    }

    /**
Return the the lock timeout set for implicit transactions, in microseconds.
<p>
This method may be called at any time during the life of the application.
<p>
@return
The the lock timeout set for implicit transactions, in microseconds.
 */
    public int getTimeout() {
        return timeout;
    }

    /* package */
    int getFlags() {
        int flags = 0;
        flags |= freeListOnly ? DbConstants.DB_FREELIST_ONLY : 0;
        flags |= freeSpace ? DbConstants.DB_FREE_SPACE : 0;

        return flags;
    }
}

File: db-4.8.30/csharp/CompactConfig.cs
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2009 Oracle.  All rights reserved.
 *
 */
using System;
using System.Collections.Generic;
using System.Text;
using BerkeleyDB.Internal;

namespace BerkeleyDB {
    /// <summary>
    /// A class to represent configuration settings for
    /// <see cref="BTreeDatabase.Compact"/> and
    /// <see cref="RecnoDatabase.Compact"/>.
    /// </summary>
    public class CompactConfig {
        private DB_COMPACT cdata;
        
        private bool fillPctSet;
        private uint fillpct;
        private bool pgsSet;
        private uint pgs;
        private bool tmoutSet;
        private uint tmout;
        /// <summary>
        /// Return the database key marking the end of the compaction operation
        /// in a Btree or Recno database. This is generally the first key of the
        /// page where the operation stopped. 
        /// </summary>
        public bool returnEnd;
        /// <summary>
        /// If non-null, the starting point for compaction. Compaction will
        /// start at the smallest key greater than or equal to
        /// <paramref name="start"/>. If null, compaction will start at the
        /// beginning of the database. 
        /// </summary>
        public DatabaseEntry start;
        /// <summary>
        /// If non-null, the stopping point for compaction. Compaction will stop
        /// at the page with the smallest key greater than
        /// <paramref name="stop"/>. If null, compaction will stop at the end of
        /// the database. 
        /// </summary>
        public DatabaseEntry stop;
        /// <summary>
        /// If true, return pages to the filesystem when possible. If false,
        /// pages emptied as a result of compaction will be placed on the free
        /// list for re-use, but never returned to the filesystem.
        /// </summary>
        /// <remarks>
        /// Note that only pages at the end of a file can be returned to the
        /// filesystem. Because of the one-pass nature of the compaction
        /// algorithm, any unemptied page near the end of the file inhibits
        /// returning pages to the file system. A repeated call to
        /// <see cref="BTreeDatabase.Compact"/> or 
        /// <see cref="RecnoDatabase.Compact"/> with a low
        /// <see cref="FillPercentage"/> may be used to return pages in this
        /// case. 
        /// </remarks>
        public bool TruncatePages;

        static internal DB_COMPACT getDB_COMPACT(CompactConfig compactData) {
            if (compactData == null)
                return null;

            if (compactData.cdata == null)
                compactData.doCompaction();
            return compactData.cdata;
        }
        
        /// <summary>
        /// Create a new CompactConfig object
        /// </summary>
        public CompactConfig() { }

        /// <summary>
        /// If non-zero, this provides the goal for filling pages, specified as
        /// a percentage between 1 and 100. Any page not at or above this
        /// percentage full will be considered for compaction. The default
        /// behavior is to consider every page for compaction, regardless of its
        /// page fill percentage. 
        /// </summary>
        public uint FillPercentage {
            get { return fillpct; }
            set {
                fillpct = value;
                fillPctSet = true;
            }
        }
        /// <summary>
        /// If non-zero, compaction will complete after the specified number of
        /// pages have been freed.
        /// </summary>
        public uint Pages {
            get { return pgs; }
            set {
                pgs = value;
                pgsSet = true;
            }
        }
        /// <summary>
        /// If non-zero, and no <see cref="Transaction"/> is specified, this
        /// parameter identifies the lock timeout used for implicit
        /// transactions, in microseconds. 
        /// </summary>
        public uint Timeout {
            get { return tmout; }
            set {
                tmout = value;
                tmoutSet = true;
            }
        }

        private void doCompaction() {
            cdata = new DB_COMPACT();

            if (fillPctSet)
                cdata.compact_fillpercent = fillpct;
            if (pgsSet)
                cdata.compact_pages = pgs;
            if (tmoutSet)
                cdata.compact_timeout = tmout;
        }

        internal uint flags {
            get {
                uint ret = 0;
                ret |= TruncatePages ? DbConstants.DB_FREE_SPACE : 0;
                return ret;
            }
        }
    }
}